connection.go

```diff
@@ -140,18 +139,53 @@ func (c *Connection) run() {
 	go c.readLoop()
 }
 
-// Close waits for pending requests to complete and then closes network
-// connection with ISO 8583 server
-func (c *Connection) Close() error {
+func (c *Connection) handleConnectionError(err error) {
+	// lock to check and update `closing`
 	c.mutex.Lock()
-	// if we are closing already, just return
-	if c.closing {
+	if err == nil || c.closing {
 		c.mutex.Unlock()
-		return nil
+		return
 	}
+
 	c.closing = true
 	c.mutex.Unlock()
 
+	// channel to wait for all goroutines to exit
+	done := make(chan bool)
+
+	c.pendingRequestsMu.Lock()
+	for _, resp := range c.respMap {
+		resp.errCh <- ErrConnectionClosed
+	}
+	c.pendingRequestsMu.Unlock()
+
+	// return error to all Send methods
+	go func() {
+		for {
+			select {
+			case req := <-c.requestsCh:
+				req.errCh <- ErrConnectionClosed
+			case <-done:
+				return
+			}
+
+		}
+	}()
+
+	go func() {
+		c.wg.Wait()
+		done <- true
+	}()
+
+	// close everything else we close normally
+	c.close()
+
+	if c.Opts.ConnectionClosedHandler != nil {
+		go c.Opts.ConnectionClosedHandler(c)
```

Comment:
He is my thoughts:
* connection is not responsible for callbacks, and any of them can block the process (`readLoop`, `writeLoop`, `Send`).
* right now, technically, there are no issues with the blocking handler because the `handleConnectionError` method and the `ConnectionClosedHandler` are the last in the chain of calls. But, it may change in the future.
* We run all callbacks in goroutines (`InboundMessageHandler`, `PingHandler`)

Reply:
#1 and #3 seem like good reasons to run this in a goroutine